docs/cli/ilab-model-serve.md

```diff
@@ -0,0 +1,101 @@
+# Design for `ilab model serve`
+
+## Background
+
+With the [request from the community](https://github.com/instructlab/instructlab/issues/1106) for `ilab` to serve different backends such as [vllm](https://docs.vllm.ai/en/stable/) and the [cli redesign](https://github.com/instructlab/dev-docs/pull/48), this design doc's purpose is to flesh out the behavior of the `ilab model serve` command.
+
+Specifically the design of the command, its flags/options, and the how a new serving backend will be tested within the project.
+
+## Design
+
+Since the subject of the `ilab model serve` command is a model, regardless of the format of the model, every command takes in the `--model` flag or uses its default value in the config.
+
+`ilab model serve` has a subcommand for each backend it can serve models with. As of this draft, the two backends `ilab` would serve with are `llama-cpp` and `vllm`.
+
+This would lead to the commands:
+
+- `ilab model serve llama-cpp`
+- `ilab model serve vllm`
+
+The flags for `ilab model serve` are either backend agnostic or specific to a subcommand and are prefixed to indicate that. The intention behind this choice is to ensure there is no mix up of flags between backends.
+
+```console
+ilab model serve
+|
+|_______ --model
+|
+|_______ (other backend agnostic flags)
+|
+|_______ llama-cpp
+|        |
+|        |_______ --llama-cpp-num-gpus
+|        |
+|        |_______ --llama-cpp-max-ctx-size
+|        |
+|        |_______ (other llama-cpp specific args)
+|
+|_______ vllm
+|        |
+|        |_______ --vllm-chat-template
+|        |
+|        |_______ (other popular vllm-specific specific args)
+|        |
+|        |_______ --vllm-args
+```
+
+## Command Examples
+
+### Bare-bones but model specific command
+
+```shell
+ilab model serve --model <PATH> 
+```
+
+- Serves the model at `<PATH>`.
+- If the `<PATH>` is the path for a `.gguf` file then `llama-cpp` is automatically used as the model serving backend.
+- If the path is a directory, `vllm` is automatically used as the backend.
+- The path provided to `vllm` can either be to a Hugging Face repo or a directory with `.safetensors`.
+
+### Bare-bones command
+
+```shell
+ilab model serve
+```
+
+- This command has the same behavior as the one above but the `--model-path` is whatever the default model path is in the config. This is the existing behavior of `ilab serve` today.
+
+### Llama-cpp backend specific commands
+
+```shell
+ilab model serve llama-cpp --model <PATH> --llama-cpp-gpu-layers 4 --llama-cpp-max-ctx-size 1024
+```
+
+- This command serves a model with `llama-cpp`.
+- If the path provided is not to a `.gguf` file, this command would error out.
```

Comment:
I don't think we should rely on the extension but have a function/script to read the image and tell us the format instead.

Reply:
I can make this statement more general, the point of the bullet is that the command auto detects that it can't serve the format specified and throws an error with feedback for the user.

Reply:
One other point to consider is that it's likely we may have multiple backends that support overlapping formats. To address that we could have a config flag to specify the intended default backend. Although that can be a later design enhancement. I only mention it to show the existing design here is compatible with expansion

Reply:
@n1hility Agreed. Great idea.